BinaryEncoding.md

```diff
@@ -436,34 +436,62 @@ The expectation is that, when a binary WebAssembly module is viewed in a browser
+| name_payload_len | `varuint32` | size of this subsection in bytes |
+| name_payload_data | `bytes` | content of this section, of length `name_payload_len` |
+
+Since name subsections have a given length, unknown or unwanted names can be
+skipped over by an engine. The current list of valid `name_type` codes are:
+
+| Name Type | Code | Description |
+| --------- | ---- | ----------- |
+| [Function](#function-names) | `1` | Assigns names to functions |
+| [Local](#local-names) | `2` | Assigns names to locals in functions |
+
+When present, name subsections must appear in this order and at most once. The
+end of the last subsection must coincide with the last byte of the name
+section to be a well-formed name section.
+
+#### Name List
+
+In any of the subsequent subsections, a `name_list` is encoded as:
 
-The sequence of `function_names` assigns names to the corresponding
-[function index](Modules.md#function-index-space). (Note: this assigns names to both
-imported and module-defined functions.) The count may differ from the actual number of functions.
+| Name Type | Type | Description |
+| --------- | ---- | ----------- |
+| count | `varuint32` | number of `name` in names |
+| names | `name*` | sequence of `name` |
+
+where a `name` is encoded as:
+
+| Name Type | Type | Description |
+| --------- | ---- | ----------- |
+| name_len | `varuint32` | number of bytes in name_str |
+| name_str | `bytes` | UTF8 encoding of the name |
 
 #### Function names
 
-| Field | Type | Description |
-| ----- | ---- | ----------- |
-| fun_name_len | `varuint32` | string length, in bytes |
-| fun_name_str | `bytes` | valid utf8 encoding |
-| local_count | `varuint32` | count of local names to follow |
-| local_names | `local_name*` | sequence of local names |
+The function names subsection is a `name_list` which assign a name to each
+function by [function index](Modules.md#function-index-space). (Note: this
+assigns names to both imported and module-defined functions.) The count
+may differ from the actual number of functions.
 
-The sequence of `local_name` assigns names to the corresponding local index. The
-count may differ from the actual number of locals.
+#### Local names
 
-#### Local name
+The local names subsection assigns a `name_list` to each function defined inside
+the module according to its index in the [Function section](#function-section).
+(Note: this does not assign names to imports; imports have no locals.) The
+`name_list` for a given function assigns a name to each local. The counts may
+differ from both the number of functions and number of locals in a given
+function.
 
 | Field | Type | Description |
 | ----- | ---- | ----------- |
-| local_name_len | `varuint32` | string length, in bytes |
-| local_name_str | `bytes` | valid utf8 encoding |
-
+| count | `varuint32` | count of `name_list` in func_locals |
+| func_locals | `name_list*` | sequence of `name_list`, one per function |
```

Comment:
This seems pretty brittle when combined with "you can avoid specifying functions if you want to". At this point why not use a map of { function_number => name } ?

I agree forcing all of them to be present is suboptimal (mixing debug and release is nice, especially if we consider transfer size). I just don't want us to end up with some super weird encoding, even if name sections aren't really standard right now.